cli/mfg.c

```diff
@@ -404,12 +412,149 @@ static int image_select(int argc, char **argv)
+		DEVICE_OPTION_NO_PAX,
+		{"img_file", .cfg_type=CFG_FILE_R, .value_addr=&cfg.fimg,
+			.argument_type=required_positional,
+			.help="firmware image file to transfer"},
+		{"yes", 'y', "", CFG_NONE, &cfg.assume_yes, no_argument,
+			"assume yes when prompted"},
+		{"force", 'f', "", CFG_NONE, &cfg.force, no_argument,
+			"force interrupting an existing fw-update command "
+			"in case firmware is stuck in the busy state"},
+		{NULL}
+	};
+
+	argconfig_parse(argc, argv, desc, opts, &cfg, sizeof(cfg));
+
+	if (switchtec_boot_phase(cfg.dev) != SWITCHTEC_BOOT_PHASE_BL1) {
+		fprintf(stderr,
+			"This command is only available in BL1 boot phase!\n");
+		fprintf(stderr,
+			"Use 'fw-update' instead to update an image in other boot phases.\n");
+		return -1;
+	}
+
+	printf("Writing the following firmware image to %s:\n",
+		switchtec_name(cfg.dev));
+
+	type = check_and_print_fw_image(fileno(cfg.fimg), cfg.img_filename);
+
+	if (type != SWITCHTEC_FW_TYPE_BL2) {
+		fprintf(stderr,
+			"Only transferring BL2 image is supported in this command.\n");
+		return -2;
+	}
+
+	ret = ask_if_sure(cfg.assume_yes);
+	if (ret) {
+		fclose(cfg.fimg);
+		return ret;
+	}
+
+	progress_start();
+	ret = switchtec_fw_write_file(cfg.dev, cfg.fimg, 1, cfg.force,
+				      progress_update);
+	fclose(cfg.fimg);
+
+	if (ret) {
+		printf("\n");
+		switchtec_fw_perror("mfg fw-transfer", ret);
+		return -3;
+	}
+
+	progress_finish();
+	printf("\n");
+
+	return 0;
+}
+
+static int fw_execute(int argc, char **argv)
+{
+	const char *desc = "Execute previously transferred firmware image (BL1 phase only)\n\n"
+			   "This command only supports executing previously "
+			   "transferred firmware when device is in BL1 boot "
+			   "phase. To activate an image in BL2 or MAIN boot "
+			   "phase, use 'fw-toggle' command instead.\n\n"
+			   BOOT_PHASE_HELP_TEXT;
+	int ret;
+
+	static struct {
+		struct switchtec_dev *dev;
+		int assume_yes;
+		enum switchtec_bl2_recovery_mode bl2_rec_mode;
+	} cfg = {
+		.bl2_rec_mode = SWITCHTEC_BL2_RECOVERY_NOT_SET
+	};
+	const struct argconfig_options opts[] = {
+		DEVICE_OPTION_NO_PAX,
+		{"yes", 'y', "", CFG_NONE, &cfg.assume_yes, no_argument,
+			"assume yes when prompted"},
+		{"bl2_recovery_mode", 'm', "MODE",
+			CFG_CHOICES, &cfg.bl2_rec_mode,
+			required_argument, "BL2 recovery mode",
+			.choices = recovery_mode_choices},
+		{NULL}
+	};
+
+	argconfig_parse(argc, argv, desc, opts, &cfg, sizeof(cfg));
+
+	if (cfg.bl2_rec_mode == SWITCHTEC_BL2_RECOVERY_NOT_SET) {
+		fprintf(stderr,
+			"BL2 recovery mode must be set in this command!\n");
+		return -1;
+	}
+
+	if (switchtec_boot_phase(cfg.dev) != SWITCHTEC_BOOT_PHASE_BL1) {
+		fprintf(stderr,
+			"This command is only available in BL1 phase!\n");
+		return -2;
+	}
+
+	if (!cfg.assume_yes)
+		printf("This command will execute the previously transferred image.\n");
+	ret = ask_if_sure(cfg.assume_yes);
+	if (ret) {
+		return ret;
+	}
+
+	ret = switchtec_fw_exec(cfg.dev, cfg.bl2_rec_mode);
```

Comment:
What happens if no image has been transferred yet? Presumably an error and I'd want that error message to be very clear but I don't see how that will happen here.